File: webrtc-sys/src/nvenc_settings.cpp
/*
 * Copyright 2025 ZenSpeak
 */

#include "livekit/nvenc_settings.h"

#include <mutex>

#include "rtc_base/logging.h"

namespace livekit_ffi {

namespace {
std::mutex g_nvenc_mutex;
NvencSettings g_nvenc_settings = NvencSettings{};
bool g_nvenc_settings_initialized = false;
}  // namespace

void SetGlobalNvencSettings(const NvencSettings& settings) {
  std::lock_guard<std::mutex> lock(g_nvenc_mutex);
  g_nvenc_settings = settings;
  g_nvenc_settings_initialized = true;
  if (settings.log_settings) {
    RTC_LOG(LS_INFO) << "Global NVENC settings set: preset="
                     << static_cast<int>(settings.preset)
                     << " profile=" << static_cast<int>(settings.profile)
                     << " rc=" << static_cast<int>(settings.rate_control)
                     << " cqp=" << settings.cqp
                     << " keyint_s=" << settings.keyframe_interval_seconds;
  }
}

NvencSettings GetGlobalNvencSettings() {
  std::lock_guard<std::mutex> lock(g_nvenc_mutex);
  if (!g_nvenc_settings_initialized) {
    // Defaults only. (We intentionally do not parse env vars here yet.)
    g_nvenc_settings_initialized = true;
  }
  return g_nvenc_settings;
}

void set_nvenc_settings_raw(int32_t preset,
                            int32_t profile,
                            int32_t rate_control,
                            int32_t cqp,
                            int32_t keyframe_interval_seconds,
                            bool log_settings) {
  NvencSettings s;
  s.preset = static_cast<NvencPreset>(preset);
  s.profile = static_cast<NvencProfile>(profile);
  s.rate_control = static_cast<NvencRateControl>(rate_control);
  s.cqp = cqp;
  s.keyframe_interval_seconds = keyframe_interval_seconds;
  s.log_settings = log_settings;
  SetGlobalNvencSettings(s);
}

}  // namespace livekit_ffi

File: webrtc-sys/src/nvenc_settings.rs
// Copyright 2025 ZenSpeak
//
// CXX bridge for process-global NVENC settings.

#[cxx::bridge(namespace = "livekit_ffi")]
pub mod ffi {
    unsafe extern "C++" {
        include!("livekit/nvenc_settings.h");

        // preset: 0=UltraLowLatency, 1=LowLatency, 2=Quality
        // profile: 0=Auto, 1=Baseline, 2=Main, 3=High
        // rate_control: 0=CBR, 1=VBR, 2=CQP
        fn set_nvenc_settings_raw(
            preset: i32,
            profile: i32,
            rate_control: i32,
            cqp: i32,
            keyframe_interval_seconds: i32,
            log_settings: bool,
        );
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[repr(i32)]
pub enum NvencPreset {
    #[default]
    UltraLowLatency = 0,
    LowLatency = 1,
    Quality = 2,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[repr(i32)]
pub enum NvencProfile {
    #[default]
    Auto = 0,
    Baseline = 1,
    Main = 2,
    High = 3,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[repr(i32)]
pub enum NvencRateControl {
    #[default]
    Cbr = 0,
    Vbr = 1,
    Cqp = 2,
}

#[derive(Debug, Clone, Copy)]
pub struct NvencSettings {
    pub preset: NvencPreset,
    pub profile: NvencProfile,
    pub rate_control: NvencRateControl,
    pub cqp: i32,
    pub keyframe_interval_seconds: i32,
    pub log_settings: bool,
}

impl Default for NvencSettings {
    fn default() -> Self {
        Self {
            // Default matches prior behavior: preset P4 + ultra-low-latency tuning.
            preset: NvencPreset::LowLatency,
            profile: NvencProfile::Auto,
            rate_control: NvencRateControl::Cbr,
            cqp: 23,
            keyframe_interval_seconds: 0,
            log_settings: true,
        }
    }
}

pub fn set_nvenc_settings(settings: NvencSettings) {
    ffi::set_nvenc_settings_raw(
        settings.preset as i32,
        settings.profile as i32,
        settings.rate_control as i32,
        settings.cqp,
        settings.keyframe_interval_seconds,
        settings.log_settings,
    );
}

